src/encoder.ts

```diff
@@ -87,136 +87,111 @@ function findTextureFormat(outputFormat: TXIOutputFormat, rle: boolean) {
   throw new Error(`No known texture format for TXI output format ${outputFormat}`);
 }
 
-function arrayBufferConcat(buffers: ArrayBuffer[]) {
-  const totalLength = buffers.reduce((acc, buf) => acc + buf.byteLength, 0);
-  const combined = new Uint8Array(totalLength);
-  let offset = 0;
-  for (const buf of buffers) {
-    combined.set(new Uint8Array(buf), offset);
-    offset += buf.byteLength;
-  }
-  return combined.buffer;
+function maxOutputSize(
+  image: ImageData,
+  outputFormat: TXIOutputFormat,
+  withRLE: boolean,
+) {
+  const { width, height } = image;
+  const bpp = textureBPP[findTextureFormat(outputFormat, withRLE)];
+
+  // RLE has no padding or duplicated rows, but in the worst case where nothing
+  // can be compressed incurs a 1-byte per pixel overhead
+  const maxBytesWithRLE = width * height * (bpp + 1);
+
+  // Without RLE, the worst case is more complex:
+  // - The final row is duplicated, so height + 1.
+  // - For each row, the final pixel is duplicated, so width + 1.
+  // - Each row is padded up to a 32-bit boundary, so add a possible
+  // 3 bytes for each row we write.
+  const maxBytesWithoutRLE = ((width + 1) * (height + 1) * bpp) + ((height + 1) * 3);
+
+  return (withRLE ? maxBytesWithRLE : maxBytesWithoutRLE) + TXI_HEADER_LENGTH;
 }
 
-class TXIEncoder {
-  private rle?: RunLengthEncoder;
-  private cursor!: BufferCursor;
-  private textureFormat!: TextureFormat;
-  private imageData!: Uint8Array;
-  private outputPixel!: Uint8Array;
-  private encoder!: PixelEncoder;
-
-  constructor(
-    private image: ImageData,
-  ) { }
-
-  private emit(pixel: Pixel) {
-    const packed = this.rle ? this.rle.encode(pixel) : pixel;
-    if (packed) this.cursor.writeArray(packed);
-  }
-
-  private process() {
-    for (let y = 0; y < this.height; y += 1) this.processRow(y);
-    if (this.rle) {
-      const leftovers = this.rle.flush();
-      if (leftovers) this.cursor.writeArray(leftovers);
-    } else {
-      this.processRow(this.height - 1);
-    }
+function encodeWithFixedRLE(
+  image: ImageData,
+  options: { outputFormat: TXIOutputFormat, rle: boolean},
+) {
+  const textureFormat = findTextureFormat(options.outputFormat, options.rle);
+  const bpp = textureBPP[textureFormat];
+  const encoder = pixelEncoders[textureFormat];
+  const rle = options.rle ? new RunLengthEncoder(bpp) : undefined;
+
+  const imageData = new Uint8Array(image.data.buffer);
+  const { width, height } = image;
+
+  const cursor = new BufferCursor(maxOutputSize(image, options.outputFormat, !!rle));
+  const outputPixel = new Uint8Array(bpp);
+  cursor.seek(TXI_HEADER_LENGTH);
+
+  function emit(pixel: Pixel) {
```

Comment:
`emit()` doesn't need an argument any more. And it can be made less branchy:
```ts
let emit: () => void;
if (rle) {
  emit = () => {
    const packed = rle.encode(outputPixel);
    if (packed) cursor.writeArray(packed);
  };
} else {
  emit = () => cursor.writeArray(outputPixel);
}
```
This may not make any difference as the JIT might be smart enough to specialize the function itself.